lane.py

```diff
@@ -387,897 +372,796 @@ def has_next_page(self):
-        or a GenreData object.
+    def _lazy_load(cls, qu, work_model):
+        """Avoid eager loading of objects that are contained in the 
+        materialized view.
         """
-        if isinstance(descriptor, tuple):
-            if len(descriptor) == 2:
-                genre, subgenres = descriptor
-            else:
-                genre, subgenres, audience_restriction = descriptor
-        else:
-            genre = descriptor
+        return qu.options(
+            lazyload(work_model.license_pool, LicensePool.data_source),
+            lazyload(work_model.license_pool, LicensePool.identifier),
+            lazyload(work_model.license_pool, LicensePool.presentation_edition),
+        )
 
-        if isinstance(genre, GenreData):
-            genredata = genre
+    @classmethod
+    def _defer_unused_fields(cls, query, work_model):
+        """Some applications use the simple OPDS entry and some
+        applications use the verbose. Whichever one we don't need,
+        we can stop from even being sent over from the
+        database.
+        """
+        if Configuration.DEFAULT_OPDS_FORMAT == "simple_opds_entry":
+            return query.options(defer(work_model.verbose_opds_entry))
         else:
-            if isinstance(genre, Genre):
-                genre_name = genre.name
-            else:
-                genre_name = genre
-            # It's in the database--just make sure it's not an old entry
-            # that shouldn't be in the database anymore.
-            genredata = classifier.genres.get(genre_name)
+            return query.options(defer(work_model.simple_opds_entry))
 
-        if not isinstance(genre, Genre):
-            genre, ignore = Genre.lookup(_db, genre)
-        return genre, genredata
 
-    @classmethod
-    def load_genres(cls, _db, genres):
-        """Turn a list of genre-like things into a list of Genre objects
-        and a list of GenreData objects.
-        """
-        genre_obj = []
-        genre_data = []
-        if genres is None:
-            return genre_obj, genre_data
-        if not any(isinstance(genres, x) for x in (list, tuple, set)):
-            genres = [genres]
-        for name in genres:
-            genre, data = cls.load_genre(_db, name)
-            genre_obj.append(genre)
-            genre_data.append(data)
-        return genre_obj, genre_data
+class LaneGenre(Base):
+    """Relationship object between Lane and Genre."""
+    __tablename__ = 'lanes_genres'
+    id = Column(Integer, primary_key=True)
+    lane_id = Column(Integer, ForeignKey('lanes.id'), index=True,
+                     nullable=False)
+    genre_id = Column(Integer, ForeignKey('genres.id'), index=True,
+                      nullable=False)
+
+    # An inclusive relationship means that books classified under the
+    # genre are included in the lane. An exclusive relationship means
+    # that books classified under the genre are excluded, even if they
+    # would otherwise be included.
```

Comment:
Does this mean that if a lane has a recursive exclusive relationship with a genre, any inclusive relationships with its subgenres will be ignored?

Reply:
That's right. I've added a test for this.